app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/EnqueuedView.kt

```diff
@@ -1,77 +1,219 @@
 			val checkmark = if (item.queueId == currentSong?.queueId) BMWRemoting.RHMIResourceIdentifier(BMWRemoting.RHMIResourceType.IMAGEID, musicImageIDs.CHECKMARK) else ""
-			return arrayOf(checkmark,
-						"",
-						UnicodeCleaner.clean(item.title ?: ""))
+
+			val coverArtImage = if (item.coverArt != null) graphicsHelpers.compress(item.coverArt!!, 90, 90, quality = 30) else ""
+
+			var title = UnicodeCleaner.clean(item.title ?: "")
+			if (title.length > ROW_LINE_MAX_LENGTH) {
+				title = title.substring(0, 20) + "..."
+			}
+			val artist = UnicodeCleaner.clean(item.artist ?: "")
+			val songMetaDataText = "${title}\n${artist}"
+
+			return arrayOf(
+					checkmark,
+					coverArtImage,
+					"",
+					songMetaDataText
+			)
 		}
 	}
 
 	init {
 		state as RHMIState.PlainState
 
 		listComponent = state.componentsList.filterIsInstance<RHMIComponent.List>().first()
+		queueImageComponent = state.componentsList.filterIsInstance<RHMIComponent.Image>().first()
+		titleLabelComponent = state.componentsList.filterIsInstance<RHMIComponent.Label>()[0]
+		subtitleLabelComponent = state.componentsList.filterIsInstance<RHMIComponent.Label>()[1]
 
 		songsEmptyList.addRow(arrayOf("", "", L.MUSIC_QUEUE_EMPTY))
 	}
 
 	fun initWidgets(playbackView: PlaybackView) {
-		state.getTextModel()?.asRaDataModel()?.value = L.MUSIC_QUEUE_TITLE
+		// this is required for pagination system
+		listComponent.setProperty(RHMIProperty.PropertyId.VALID, false)
+
 		listComponent.setVisible(true)
-		listComponent.setProperty(RHMIProperty.PropertyId.LIST_COLUMNWIDTH, "57,50,*")
+		listComponent.setProperty(RHMIProperty.PropertyId.LIST_COLUMNWIDTH, "57,90,10,*")
 		listComponent.getAction()?.asHMIAction()?.getTargetModel()?.asRaIntModel()?.value = playbackView.state.id
 		listComponent.getAction()?.asRAAction()?.rhmiActionCallback = RHMIActionListCallback { onClick(it) }
+		listComponent.getSelectAction()?.asRAAction()?.rhmiActionCallback = RHMIActionListCallback { onSelectAction(it) }
 	}
 
 	fun show() {
 		currentSong = musicController.getMetadata()
+		val newQueueMetadata = musicController.getQueue()
+
+		// same queue as before, just select currently playing song
+		if (queueMetadata != null && queueMetadata == newQueueMetadata) {
+			showCurrentlyPlayingSong()
+			return
+		}
+
+		queueMetadata = newQueueMetadata
 		songsList.clear()
-		val songs = musicController.getQueue()
+		val songs = queueMetadata?.songs
 		if (songs?.isNotEmpty() == true) {
 			listComponent.setEnabled(true)
 			listComponent.setSelectable(true)
 			songsList.addAll(songs)
-			listComponent.getModel()?.setValue(songsListAdapter, 0, songsListAdapter.height, songsListAdapter.height)
-
-			// set the selection to the current song
-			val index = songs.indexOfFirst { it.queueId == currentSong?.queueId }
-			if (index >= 0) {
-				state.app.events.values.firstOrNull { it is RHMIEvent.FocusEvent }?.triggerEvent(
-						mapOf(0.toByte() to listComponent.id, 41.toByte() to index)
-				)
+
```

Comment:
I found a bug with VLC, which had the symptom of an empty list: It couldn't find where to draw the currently-selected checkbox, so it never sent an initial list to the car, so the car never sent a RequestDataCallback. I was able to fix it by adding a `showList()` here, to get a basic list to start, and then rely on the RequestDataCallback to fill in the rest.

Reply:
Oh interesting, so you added a showList right before the listComponent requestDataCallback assignment then? I added that call there though

Reply:
Indeed, I wound up needing it outside the RequestDataCallback as well. I think it's because the car has no idea that the list has any data in it, until some sort of showData call happens to set the list's max size first.

Reply:
That makes sense, I added the showList call just before the listComponent.requestDataCallback assignment so that should fix this problem